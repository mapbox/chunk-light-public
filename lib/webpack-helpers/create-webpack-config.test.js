'use strict';

const path = require('path');
const tempy = require('tempy');
const fs = require('fs');

const getDefaultConfig = require('../default-underreact.config');
const Urc = require('../config/urc');
const createConfig = require('./create-webpack-config');

test('Basic Test', () => {
  const defaultConfig = getDefaultConfig({
    mode: 'development',
    configPath: path.join(process.cwd(), 'underreact.config.js')
  });

  const urc = new Urc(
    {
      polyfills: {
        promise: true,
        fetch: true
      },
      siteBasePath: 'fancy',
      publicAssetsPath: 'cacheable-things',
      stylesheets: [path.join(__dirname, './src/bg.css')],
      htmlSource: () => {
        return `<html></html>`;
      },
      webpackLoaders: [
        {
          test: /\.less$/,
          use: ['css-loader', 'less-loader']
        }
      ],
      webpackPlugins: [function foo() {}],
      webpackConfigTransform: config => {
        config.devtool = false;
        return config;
      }
    },
    defaultConfig
  );
  expect(createConfig(urc)).toMatchSnapshot();
});

test('Uses babelrc if it exists at project root', () => {
  const tempDir = tempy.directory();
  const defaultConfig = getDefaultConfig({
    mode: 'development',
    configPath: path.join(tempDir, 'underreact.config.js')
  });
  const urc = new Urc({}, defaultConfig);

  fs.writeFileSync(path.join(tempDir, '.babelrc'), `{}`);
  expect(
    createConfig(urc).module.rules.find(
      obj => obj.test.toString() === `/\\.jsx?$/`
    ).use
  ).toMatchObject([
    {
      loader: require.resolve('babel-loader'),
      options: {
        cacheDirectory: true,
        babelrc: true,
        compact: false
      }
    }
  ]);
});

test('Uses babel preset if babelrc doesnt exists at project root', () => {
  const tempDir = tempy.directory();
  const defaultConfig = getDefaultConfig({
    mode: 'development',
    configPath: path.join(tempDir, 'underreact.config.js')
  });
  const urc = new Urc({}, defaultConfig);

  expect(
    createConfig(urc).module.rules.find(
      obj => obj.test.toString() === `/\\.jsx?$/`
    ).use
  ).toMatchObject([
    {
      loader: require.resolve('babel-loader'),
      options: {
        presets: [require.resolve('../../packages/babel-preset-mapbox')],
        cacheDirectory: true,
        babelrc: false,
        compact: false
      }
    }
  ]);
});

test('babel-loader creates a good cacheIdentifier which includes urc.browserslist', () => {
  const defaultConfig = getDefaultConfig({
    mode: 'development',
    configPath: '/fake/underreact.config.js'
  });

  const browserslist = 'last 2 versions';
  const urc = new Urc({ browserslist: [browserslist] }, defaultConfig);

  expect(
    createConfig(urc).module.rules.find(
      obj => obj.test.toString() === `/\\.jsx?$/`
    ).use
  ).toMatchObject([
    {
      loader: require.resolve('babel-loader'),
      options: {
        cacheDirectory: true,
        babelrc: false,
        compact: false,
        cacheIdentifier: expect.stringMatching(browserslist)
      }
    }
  ]);
});

test('config transform works', () => {
  const defaultConfig = getDefaultConfig({
    mode: 'development',
    configPath: path.join(process.cwd(), 'underreact.config.js')
  });
  const urc = new Urc(
    {
      webpackConfigTransform: obj =>
        Object.assign({ injectedObj: 'injectedObj' }, obj)
    },
    defaultConfig
  );
  expect(createConfig(urc)).toMatchObject({
    injectedObj: 'injectedObj'
  });
});
