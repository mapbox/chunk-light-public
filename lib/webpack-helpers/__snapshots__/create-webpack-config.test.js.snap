// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Test 1`] = `
Object {
  "bail": false,
  "devtool": false,
  "entry": Object {
    "main": Array [
      "<PROJECT_ROOT>/src/entry.js",
    ],
    "polyfill": "<PROJECT_ROOT>/polyfill/index.js",
    "vendor": Array [
      "<PROJECT_ROOT>/node_modules/react",
      "<PROJECT_ROOT>/node_modules/react-dom",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "exclude": Array [
              /\\[\\\\/\\\\\\\\\\\\\\\\\\]node_modules\\[\\\\/\\\\\\\\\\\\\\\\\\]/,
            ],
            "test": /\\\\\\.jsx\\?\\$/,
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/babel-loader/lib/index.js",
                "options": Object {
                  "babelrc": false,
                  "cacheCompression": false,
                  "cacheDirectory": true,
                  "cacheIdentifier": "{\\"babel-loader\\":\\"8.0.2\\",\\"babel-core\\":\\"6.26.3\\",\\"babelrc\\":\\"\\",\\"env\\":\\"test\\"}",
                  "compact": false,
                  "presets": Array [
                    [Function],
                  ],
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "<PROJECT_ROOT>/node_modules/style-loader/index.js",
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/postcss-loader/src/index.js",
                "options": Object {
                  "ident": "postcss",
                  "plugins": Array [
                    [Function],
                  ],
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.less\\$/,
            "use": Array [
              "css-loader",
              "less-loader",
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(js\\|jsx\\)\\$/,
              /\\\\\\.template\\$/,
              /\\\\\\.json\\$/,
              /\\\\\\.ejs\\$/,
            ],
            "loader": "<PROJECT_ROOT>/node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "name": "cacheable-things/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
    "strictExportPresence": true,
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "minimizer": Array [
      UglifyJsPlugin {
        "options": Object {
          "cache": true,
          "exclude": undefined,
          "extractComments": false,
          "include": undefined,
          "parallel": true,
          "sourceMap": true,
          "test": /\\\\\\.js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "uglifyOptions": Object {
            "compress": Object {
              "comparisons": false,
              "ecma": 5,
              "warnings": false,
            },
            "mangle": Object {
              "safari10": true,
            },
            "output": Object {
              "ascii_only": true,
              "comments": false,
              "ecma": 5,
            },
            "parse": Object {
              "ecma": 8,
            },
          },
          "warningsFilter": [Function],
        },
      },
    ],
    "runtimeChunk": "single",
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "initial",
          "enforce": true,
          "name": "vendor",
          "test": "vendor",
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "cacheable-things/js/[name].chunk.js",
    "filename": "cacheable-things/js/[name].js",
    "path": "<PROJECT_ROOT>/_underreact-site",
    "publicPath": "/fancy/",
  },
  "performance": false,
  "plugins": Array [
    ModuleNotFoundPlugin {
      "appPath": "<PROJECT_ROOT>",
      "getRelativePath": [Function],
      "prettierError": [Function],
    },
    AssetsWebpackPlugin {
      "options": Object {
        "filename": "assets.json",
        "fullPath": true,
        "manifestFirst": true,
        "path": "<PROJECT_ROOT>/_underreact-site",
        "prettyPrint": false,
        "processOutput": [Function],
        "update": false,
      },
      "writer": [Function],
    },
    EnvironmentPlugin {
      "defaultValues": Object {
        "DEPLOY_ENV": "test",
        "NODE_ENV": "test",
      },
      "keys": Array [
        "NODE_ENV",
        "DEPLOY_ENV",
      ],
    },
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "cacheable-things/css/[name]-[contenthash:8].css",
        "filename": "cacheable-things/css/[name]-[contenthash:8].css",
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [
          "polyfill",
          "runtime",
        ],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": Object {},
        "minify": Object {},
        "showErrors": true,
        "template": "<PROJECT_ROOT>/node_modules/html-webpack-plugin/default_index.ejs",
        "templateContent": [Function],
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    [Function],
    ExcludeChunksPlugin {
      "excludeChunks": Array [
        "runtime",
      ],
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@babel/runtime": "<PROJECT_ROOT>/node_modules/@babel/runtime",
    },
  },
  "resolveLoader": Object {
    "modules": Array [
      "<PROJECT_ROOT>/node_modules",
      "node_modules",
    ],
  },
}
`;
